pkg/genericapiserver/server/serve.go

```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	return nil
+}
+
+func (ln tcpKeepAliveHandler) wrap(conn net.Conn) (net.Conn, error) {
+	return conn, nil
 }
 
-func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
-	tc, err := ln.AcceptTCP()
+// ipBasedLimitListener accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
+//
+// Based on Go 1.7.4 net/netutil/listen.go
+// TODO: (cheftako) : IPv6, CIDR Range rather than individual IP Addresses?
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+type ipBasedLimit struct {
+	// Map in which we map the serialized ip to the current count
+	limits map[string]int
+
+	// Current total count of the connections in existance (no localhost)
+	total int
+
+	// Lock used to ensure the limits and total are updated atomicaly
+	mutex sync.Mutex
+
+	// The maximum concurent connections an IP is allowed to have open
+	ipMax int
+
+	// The total maximum concurrent connections allowed open (Except localhost)
+	max int
+}
+
+func (i *ipBasedLimit) incrementIp(ipser string) error {
+	if i.max <= 0 {
+		return nil
+	}
+	ip := net.ParseIP(ipser)
+	if ip == nil {
+		glog.Infof("Received request from client with invalid ip %s", ipser)
+		return limitedError{ip: ipser}
+	}
+	if ip.IsLoopback() {
+		return nil
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	current := i.limits[ipser]
+	if current >= i.ipMax || i.total >= i.max {
+		return limitedError{ip: ipser}
+	}
+	i.total++
+	i.limits[ipser] = current + 1
+	return nil
+}
+
+func (i *ipBasedLimit) decrementIp(ipser string) {
+	if i.max <= 0 {
+		return
+	}
+	ip := net.ParseIP(ipser)
+	if ip == nil || ip.IsLoopback() {
+		return
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	i.total--
+	current := i.limits[ipser]
+	if current > 1 {
+		i.limits[ipser] = current - 1
+	} else {
+		glog.Infof("Cleaned up all references for IP %s", ipser)
+		delete(i.limits, ipser)
+	}
+}
+
+// The actual structure/handler which attaches the limiting code to a connection
+type ipBasedLimitHandler struct {
+	// The limit structure for this connection type.
+	ipLimit *ipBasedLimit
+}
+
+func (h ipBasedLimitHandler) init(conn *net.TCPConn) error {
+	return nil
+}
+
+func (h ipBasedLimitHandler) wrap(conn net.Conn) (net.Conn, error) {
+	ip, _, err := net.SplitHostPort(conn.RemoteAddr().String())
 	if err != nil {
+		conn.Close() // Not decrementing as we have not yet incremented
+		glog.Infof("Something went wrong getting the remote address %v", err)
 		return nil, err
 	}
-	tc.SetKeepAlive(true)
-	tc.SetKeepAlivePeriod(defaultKeepAlivePeriod)
-	return tc, nil
+	local := conn.LocalAddr().String()
+	glog.Infof("Entered ip based limit Accept() for IP %s on %s", ip, local)
+	err = h.ipLimit.incrementIp(ip)
+	if err != nil {
+		conn.Close() // Not decrementing as we did not successfully increment
+		glog.Infof("Denied accept to IP %s", ip)
```

Comment:
The goal here is to limit concurrent connection prior to taking on the expense of having completed (or even started) the SSL handshake. Given that I don't think there is a practical way to return a usable HTTP or API response.